content/src/content/jcr_root/apps/core/wcm/components/image/v3/image/clientlibs/site/js/image.js

```diff
@@ -0,0 +1,291 @@
+
+    function getOptimalWidth(sizes, width) {
+        var len = sizes.length;
+        var key = 0;
+
+        while ((key < len - 1) && (sizes[key] < width)) {
+            key++;
+        }
+
+        return sizes[key].toString();
+    }
+
+    var triggerEvent = function(elem, name, detail) {
+        var event = document.createEvent("Event");
+
+        if (!detail) {
+            detail = {};
+        }
+        detail.instance = cmpImage;
+        event.initEvent(name, true, true);
+        event.detail = detail;
+        elem.dispatchEvent(event);
+        return event;
+    };
+
+    var debounce = function(func) {
+        var timeout;
+        var timestamp;
+        var wait = 99;
+        var run = function() {
+            timeout = null;
+            func();
+        };
+        var later = function() {
+            var last = Date.now() - timestamp;
+
+            if (last < wait) {
+                setTimeout(later, wait - last);
+            } else {
+                (requestIdleCallback || run)(run);
+            }
+        };
+
+        return function() {
+            timestamp = Date.now();
+
+            if (!timeout) {
+                timeout = setTimeout(later, wait);
+            }
+        };
+    };
+
+    var getAutoSmartCrops = function(src) {
+        var autoSmartCrops = {};
+        var request = new XMLHttpRequest();
+        var url = decodeURIComponent(src).split(SRC_URI_TEMPLATE_WIDTH_VAR)[0] + "?req=set,json";
+
+        request.open("GET", url, false);
+        request.onload = function() {
+            if (request.status >= 200 && request.status < 400) {
+                // success status
+                var responseText = request.responseText;
+                var rePayload = new RegExp(/^(?:\/\*jsonp\*\/)?\s*([^()]+)\(([\s\S]+),\s*"[0-9]*"\);?$/gmi);
+                var rePayloadJSON = new RegExp(/^{[\s\S]*}$/gmi);
+                var resPayload = rePayload.exec(responseText);
+                var payload;
+                if (resPayload) {
+                    var payloadStr = resPayload[2];
+                    if (rePayloadJSON.test(payloadStr)) {
+                        payload = JSON.parse(payloadStr);
+                    }
+
+                }
+                // check "relation" - only in case of smartcrop preset
+                if (payload && payload.set.relation && payload.set.relation.length > 0) {
+                    for (var i = 0; i < payload.set.relation.length; i++) {
+                        autoSmartCrops[parseInt(payload.set.relation[i].userdata.SmartCropWidth)] =
+                            ":" + payload.set.relation[i].userdata.SmartCropDef;
+                    }
+                }
+            } else {
+                // error status
+            }
+        };
+        request.send();
+        return autoSmartCrops;
+    };
+
+    var image = (function() {
+        var images;
+
+        var setSizes = function(imgEl, width) {
+            imgEl._autoWidth = width;
```

Comment:
This code needs some documentation, it's hard to follow the logic. Also, is it even safe to use `_autoWidth`?